filament/src/MorphTargetBuffer.cpp

```diff
@@ -74,114 +74,110 @@ static inline size_t getHeight(size_t vertexCount) noexcept {
 }
 
 void FMorphTargetBuffer::setPositionsAt(FEngine& engine, size_t targetIndex, math::float3 const* positions, size_t count) {
     ASSERT_PRECONDITION(targetIndex < mCount, "targetIndex must be < count");
 
-    auto size = getSize(mVertexCount);
+    auto size = getSize<VertexAttribute::POSITION>(mVertexCount);
 
-    ASSERT_PRECONDITION((int)sizeof(math::float3) * count <= size,
-            "MorphTargetBuffer (size=%lu) overflow (size=%lu)",
-            size, sizeof(math::float3) * count);
+    ASSERT_PRECONDITION(sizeof(float3) * count <= size,
+            "MorphTargetBuffer (size=%lu) overflow (size=%lu)", size, sizeof(float3) * count);
 
     // We could use a pool instead of malloc() directly.
-    auto* out = (float4*) malloc(size);
-    std::transform(positions, positions + count, out,
-            [](const float3& p) { return float4(p, 1.0f); });
+    auto* out = (float3*) malloc(size);
+    memcpy(out, positions, sizeof(float3) * count);
 
-    Texture::PixelBufferDescriptor buffer(out, size, Texture::Format::RGBA, Texture::Type::FLOAT,
+    Texture::PixelBufferDescriptor buffer(out, size, Texture::Format::RGB, Texture::Type::FLOAT,
             FREE_CALLBACK);
-
     FEngine::DriverApi& driver = engine.getDriverApi();
-    driver.update3DImage(mTbHandle, 0, 0, 0, getPositionIndex(targetIndex),
+    driver.update3DImage(mPbHandle, 0, 0, 0, targetIndex,
             getWidth(mVertexCount), getHeight(mVertexCount), 1,
             std::move(buffer));
 }
 
 void FMorphTargetBuffer::setPositionsAt(FEngine& engine, size_t targetIndex, math::float4 const* positions, size_t count) {
     ASSERT_PRECONDITION(targetIndex < mCount, "targetIndex must be < count");
 
-    auto size = getSize(mVertexCount);
+    auto size = getSize<VertexAttribute::POSITION>(mVertexCount);
 
-    ASSERT_PRECONDITION((int)sizeof(math::float4) * count <= size,
-            "MorphTargetBuffer (size=%lu) overflow (size=%lu)",
-            size, sizeof(math::float4) * count);
+    ASSERT_PRECONDITION(sizeof(float3) * count <= size,
+            "MorphTargetBuffer (size=%lu) overflow (size=%lu)", size, sizeof(float3) * count);
 
     // We could use a pool instead of malloc() directly.
-    auto* out = (float4*) malloc(size);
-    memcpy(out, positions, sizeof(math::float4) * count);
+    auto* out = (float3*) malloc(size);
+    std::transform(positions, positions + count, out,
+            [](const float4& p) { return float3(p.xyz); });
 
-    Texture::PixelBufferDescriptor buffer(out, size, Texture::Format::RGBA, Texture::Type::FLOAT,
+    Texture::PixelBufferDescriptor buffer(out, size, Texture::Format::RGB, Texture::Type::FLOAT,
             FREE_CALLBACK);
-
     FEngine::DriverApi& driver = engine.getDriverApi();
-    driver.update3DImage(mTbHandle, 0, 0, 0, getPositionIndex(targetIndex),
+    driver.update3DImage(mPbHandle, 0, 0, 0, targetIndex,
             getWidth(mVertexCount), getHeight(mVertexCount), 1,
             std::move(buffer));
```

Comment:
you could factor that code out, probably, because it's identical with the float4 version. Might save a few bytes of code.

Reply:
Ok it's done.